tools/migration.py

```diff
@@ -0,0 +1,248 @@
+    """ Class for migrating tasks between two instances.
+    Tasks will only be migrated if all connected findings are on the new instance as well.
+    """
+
+    def migrate(self):
+        """ Migrates the tasks. """
+        parameters = {"with-count": True, "max": 0}
+        new_tasks_count = self.get_from_new("tasks", parameters=parameters)["totalTaskCount"] + 1
+
+        old_tasks = self.get_from_old("tasks", parameters={"details": True})
+        total = len(old_tasks)
+        for old_task in old_tasks:
+            if self.task_exists(old_task):
+                total -= 1
+            else:
+                if self.adjust_task(old_task, new_tasks_count):
+                    print("migrating task %s" % old_task["id"])
+                    new_tasks_count += 1
+                    self.add_task(old_task)
+        if total == 0:
+            print("No new tasks to migrate.")
+        else:
+            print("migrated %d/%d tasks" % (self.migrated, total))
+
+    def adjust_task(self, task, new_id):
+        """ Before adding the task to the new instance, it needs to get a new id, in order to prevent
+        that potential existing tasks are overwritten. The ids of any connected findings need
+        to be changed to the corresponding findings on the new instance, as well.
+        If any finding cannot be matched on the new instance `False` will be returned, `True` otherwise."""
+        for finding in task["findings"]:
+            matching_finding_id = self.get_matching_finding_id(finding["findingId"])
+            if not matching_finding_id:
+                print("The finding %s for the task %s does not exists on the new instance." % (
+                    finding["findingId"], task["id"]))
+                return False
+            finding["findingId"] = matching_finding_id
+        task["id"] = new_id
+
+        return True
+
+    def add_task(self, task):
+        """ Adds a task to the new instance """
+        self.migrated += 1
+        self.put("tasks", path_param=str(task["id"]), data=task)
+
+    def task_exists(self, old_task):
+        """ Checks if the given tasks already exists on the new instance. """
+        new_tasks = self.get_from_new("tasks", parameters={
+            "author": old_task["author"],
+            "assignee": old_task["assignee"],
+            "tags": old_task["tags"],
+            "details": True
+        })
+
+        for new_task in new_tasks:
+            if self.superficial_match(new_task, old_task) and self.task_findings_match(new_task, old_task):
+                return True
+        return False
+
+    def task_findings_match(self, new, old):
+        """ Checks if the findings of the given two tasks are the same. """
+        new_findings = self.get_task_findings(self.new, new)
+        old_findings = self.get_task_findings(self.old, old)
+
+        for old_finding in old_findings:
+            match_found = False
+            for new_finding in new_findings:
```

Comment:
extract a method for this loop? that way, you can get rid of the match_found variable